File: from_scratch/reconstruction.py
import numpy as np

# Définir la taille N
N = 6

# Fonction pour charger les points d'image
def load_points_image(file_name):
    points = np.loadtxt(file_name)
    return points

# Fonction pour charger les points réels
def load_points_real(file_name):
    points = np.loadtxt(file_name)
    return points


# Fonction pour lire la matrice de projection P à partir du fichier
def load_projection_matrix(file_name):
    # Lire les dimensions de la matrice P
    with open(file_name, 'r') as f:
        rows, cols = map(int, f.readline().split())  # Lire les dimensions 3 x 4
        P = np.loadtxt(f)  # Lire les éléments de la matrice
    return P

def solve_library(A, Y):
    # Résoudre le système A * X = Y avec lstsq
    X, residuals, rank, s = np.linalg.lstsq(A, Y, rcond=None)
    return X


def solve_with_svd(A, Y):
    # Dimensions de la matrice A
    n, m = A.shape

    # Effectuer la décomposition SVD
    U, S, Vt = np.linalg.svd(A)

    # Créer la matrice S de taille n * m
    S_full = np.zeros((n, m))  # Matrice S de taille n x m
    np.fill_diagonal(S_full, S)  # Remplir la diagonale avec les valeurs singulières

    # Calcul de Σ^+ (pseudo-inverse de la matrice S)
    S_plus = np.zeros((m, n))  # S+ sera de taille m x n
    for i in range(min(n, m)):
        if abs(S[i]) > 1e-10:  # Vérifier les valeurs singulières non nulles
            S_plus[i, i] = 1.0 / S[i]
    # Calcul de A^+ = V * Σ^+ * U^T
    A_plus = np.dot(np.dot(Vt.T, S_plus), U.T)

    # Calcul de la solution X = A^+ * Y
    solution = np.dot(A_plus, Y)

    return solution


# Fonction principale
def main(image_name):
    # Initialisation des tableaux pour les points
    X = np.zeros(N)
    Y = np.zeros(N)
    Z = np.zeros(N)
    u = np.zeros(N)
    v = np.zeros(N)

    # Construction des noms de fichiers
    points_image_file = f"points/points_{image_name}.txt"
    points_reel_file = f"points/points_reel{image_name}.txt"
    p_file = f"P-{image_name}.txt"
    output_file = f"points/points_reconstruit{image_name}.txt"

    print("Lecture des points ...")
    u, v = load_points_image(points_image_file).T
    X, Y, Z = load_points_real(points_reel_file).T

    print("Lecture de P ...")
    P = load_projection_matrix(p_file)

    print("Reconstruction ...")

    # Ouvrir le fichier de sortie
    with open(output_file, "w") as output:
        for i in range(N):
            point_image = np.array([u[i], v[i], 1.0])  # Coordonnée homogène du point d'image
            print(f"Calcul pour le point {i} : ")
            print(point_image)

            # Résoudre pour X (les coordonnées réelles)
            result = solve_library(P, point_image)

            # Normalisation des coordonnées en divisant par la dernière composante (coordonnée homogène)
            result /= result[3]

            print(f"Coordonnées : {result[0]} {result[1]} {result[2]}")
            # Écriture des coordonnées dans le fichier
            output.write(f"{result[0]} {result[1]} {result[2]}\n")

    print(f"Coordonnées reconstruites écrites dans : {output_file}")

if __name__ == "__main__":
    # Exemple d'utilisation avec le nom de l'image passé en argument
    image_name = "cube"  # Remplacer par le nom de l'image voulu
    main(image_name)

File: from_scratch/reconstruction.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <math.h>
#include "matrice.h"
#include "manipulation_fichiers.h"
#include "SVD.h"
#define N 6

// Fonction pour résoudre le système A*X=Y à l'aide de SVD
matrice resolution_systeme_SVD(matrice* A, matrice* Y) {
    // Dimensions de la matrice A
    int n = A->n;
    int m = A->m;

    // Matrices pour la décomposition SVD
    matrice* U = matrice_nulle_pointeur(n, n);   // Matrice orthogonale U (n x n)
    matrice* S = matrice_nulle_pointeur(n, m);   // Matrice diagonale Σ (n x m)
    matrice* V = matrice_nulle_pointeur(m, m);  // Matrice orthogonale V^T (m x m)

    // Effectuer la décomposition SVD
    qr_algorithm_SVD(A, U, S, V);

    // Calcul de Σ^+ (pseudo-inverse de la matrice Σ)
    matrice* S_plus = matrice_nulle_pointeur(m, n);  // Σ^+ (m x n)
    for (int i = 0; i < n && i < m; i++) {
        long double sigma = S->mat[i][i];
        if (fabs(sigma) > 1e-10) {  // Éviter les divisions par zéro
            S_plus->mat[i][i] = 1.0 / sigma;
        }
    }

    // Calcul de A^+ = V * Σ^+ * U^T

    matrice V_S_plus = produit(*V, *S_plus); // Stocke V * Σ^+
    matrice U_transpose = transposee(*U);
    matrice A_plus = produit(V_S_plus, U_transpose);

    // Calcul de la solution X = A^+ * Y
    matrice solution = produit(A_plus, *Y);
    // Libération des matrices
    free_matrice(*U);
    free_matrice(*S);
    free_matrice(*V);
    free_matrice(*S_plus);
    free_matrice(V_S_plus);
    free_matrice(U_transpose);
    free_matrice(A_plus);

    // Retourner la solution
    return solution;
}

int main(int argc, char const *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <nom_de_l'image>\n", argv[0]);
        return 1;
    }
    long double* X = calloc(N, sizeof(long double));
    long double* Y = calloc(N, sizeof(long double));
    long double* Z = calloc(N, sizeof(long double));
    long double* u = calloc(N, sizeof(long double));
    long double* v = calloc(N, sizeof(long double));

    // Récupérer le nom de l'image depuis les arguments
    const char* image_name = argv[1];
    char points_image_file[256];
    char points_reel_file[256];
    snprintf(points_image_file, sizeof(points_image_file), "points/points_%s.txt", image_name);
    snprintf(points_reel_file, sizeof(points_reel_file), "points/points_reel%s.txt", image_name);

    printf("Lecture des points ...\n");
    load_all_points_images(points_image_file, u, v);
    load_all_points_reels(points_reel_file, X, Y, Z);

    char p_file[256];
    snprintf(p_file, sizeof(p_file), "P-%s.txt", image_name);
    printf("Lecture P ...\n");
    matrice P = lire_matrice(p_file);
    print_matrice(P);

    printf("Reconstruction ...\n");

    // Préparer le fichier de sortie
    char output_file[256];
    snprintf(output_file, sizeof(output_file), "points/points_reconstruit%s.txt", image_name);
    FILE* output = fopen(output_file, "w");
    if (!output) {
        perror("Erreur lors de l'ouverture du fichier de sortie");
        return 1;
    }

    for (int i = 0; i < N; ++i) {
        matrice point = matrice_nulle(3, 1);
        point.mat[0][0] = u[i];
        point.mat[1][0] = v[i];
        point.mat[2][0] = 1.0;
        printf("calcul pour le point %d : \n", i);
        print_matrice(point);
        matrice result = resolution_systeme_SVD(&P, &point);

        // Normalisation des coordonnées
        for (int j = 0; j < 4; ++j) {
            result.mat[j][0] /= result.mat[3][0];  // Normalisation en divisant par la dernière composante
        }
        printf("Coordonnées : %Lf %Lf %Lf\n", result.mat[0][0], result.mat[1][0], result.mat[2][0]);
        // Écriture des coordonnées dans le fichier
        fprintf(output, "%Lf %Lf %Lf\n", result.mat[0][0], result.mat[1][0], result.mat[2][0]);

        // Libération des matrices allouées
        free_matrice(point);
        free_matrice(result);
    }

    fclose(output);
    printf("Coordonnées reconstruites écrites dans : %s\n", output_file);

    // Libération des tableaux
    free(X);
    free(Y);
    free(Z);
    free(u);
    free(v);

    return 0;
}
